odl/solvers/functional/default_functionals.py

```diff
@@ -0,0 +1,469 @@
+        evaluating the gradient operator in this point it will return 0.
+        """
+        functional = self
+
+        class L1Gradient(Operator):
+
+            """The gradient operator of this functional."""
+
+            def __init__(self):
+                """Initialize a new instance."""
+                super().__init__(functional.domain, functional.domain,
+                                 linear=False)
+
+            def _call(self, x):
+                """Apply the gradient operator to the given point."""
+                return x.ufunc.sign()
+
+        return L1Gradient()
+
+    @property
+    def proximal(self):
+        """Return the proximal factory of the functional.
+
+        See Also
+        -----
+        proximal_l1 : proximal factory for the L1-norm.
+        """
+        return proximal_l1(space=self.domain)
+
+    @property
+    def convex_conj(self):
+        """The convex conjugate functional of the L1-norm."""
+        return IndicatorLpUnitBall(self.domain, exponent=np.inf)
+
+
+class IndicatorLpUnitBall(Functional):
+
+    """The indicator function on the unit ball in given the ``Lp`` norm.
+
+    It does not implement `gradient` since it is not differentiable everywhere.
+
+    Notes
+    -----
+    This functional is defined as
+
+        .. math::
+
+            f(x) = \\left\{ \\begin{array}{ll}
+            0 & \\text{if } ||x||_{L_p} \\leq 1, \\\\
+            \\infty & \\text{else,}
+            \\end{array} \\right.
+
+    where :math:`||x||_{L_p}` is the :math:`L_p`-norm, which for finite values
+    of :math:`p` is defined as
+
+        .. math::
+
+            \| x \|_{L_p} = \\left( \\int_{\Omega} |x|^p dx \\right)^{1/p},
+
+    and for :math:`p = \\infty` it is defined as
+
+        .. math::
+
+            ||x||_{\\infty} = \max_x (|x|).
+
+    The functional also allows noninteger and nonpositive values of the
+    exponent :math:`p`, however in this case :math:`\| x \|_{L_p}` is not a
+    norm.
+    """
+
+    def __init__(self, space, exponent):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : `DiscreteLp` or `FnBase`
+            Domain of the functional.
+        exponent : int or infinity
+            Specifies wich norm to use.
+        """
+        super().__init__(space=space, linear=False)
+        self.__exponent = float(exponent)
+
+    @property
+    def exponent(self):
+        """Exponent corresponding to the norm."""
+        return self.__exponent
+
+    # TODO: update when integration operator is in place: issue #440
+    def _call(self, x):
+        """Apply the functional to the given point."""
+        if self.exponent == 1:
+            x_norm = x.ufunc.absolute().inner(self.domain.one())
+        elif self.exponent == 2:
+            x_norm = x.norm()
+        elif self.exponent == np.inf:
+            x_norm = x.ufunc.absolute().ufunc.max()
+        else:
+            x.ufunc.absolute(out=x)
```

Comment:
This modifies `x` in place. Which we require that `_call` does not (since it is given a reference).